piecrust/src/store/tree.rs

```diff
@@ -0,0 +1,151 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use bytecheck::CheckBytes;
+use piecrust_uplink::ModuleId;
+use rkyv::{Archive, Deserialize, Serialize};
+
+// This means we have max `2^32` modules
+const HEIGHT: usize = 32;
+const ARITY: usize = 2;
+
+pub type Tree = dusk_merkle::Tree<Hash, HEIGHT, ARITY>;
+
+#[derive(
+    Debug,
+    Clone,
+    Copy,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    Hash,
+    Archive,
+    Deserialize,
+    Serialize,
+    CheckBytes,
+)]
+#[archive(as = "Self")]
+pub struct Hash([u8; blake3::OUT_LEN]);
+
+impl Hash {
+    pub fn as_bytes(&self) -> &[u8; blake3::OUT_LEN] {
+        &self.0
+    }
+}
+
+impl From<Hash> for [u8; blake3::OUT_LEN] {
+    fn from(hash: Hash) -> Self {
+        hash.0
+    }
+}
+
+impl From<[u8; blake3::OUT_LEN]> for Hash {
+    fn from(bytes: [u8; blake3::OUT_LEN]) -> Self {
+        Self(bytes)
+    }
+}
+
+impl AsRef<[u8]> for Hash {
+    fn as_ref(&self) -> &[u8] {
+        &self.0
+    }
+}
+
+impl dusk_merkle::Aggregate<HEIGHT, ARITY> for Hash {
+    const EMPTY_SUBTREES: [Hash; HEIGHT] = EMPTY_SUBTREES;
+
+    fn aggregate<'a, I>(items: I) -> Self
+    where
+        Self: 'a,
+        I: Iterator<Item = &'a Self>,
+    {
+        let mut hasher = Hasher::new();
+        for item in items {
+            hasher.update(item.as_bytes());
+        }
+        hasher.finalize()
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Hasher(blake3::Hasher);
+
+impl Hasher {
+    pub fn new() -> Self {
+        Self(blake3::Hasher::new())
+    }
+
+    pub fn update(&mut self, input: &[u8]) -> &mut Self {
+        self.0.update(input);
+        self
+    }
+
+    pub fn finalize(&self) -> Hash {
+        let hash = self.0.finalize();
+        Hash(hash.into())
+    }
+}
```

Comment:
What's the reason behind the creation of this `Hasher` type? Why not use `blake3::Hasher` directly?

Reply:
In the future it would be nice to be able to change hash function by just targeting a single file. I should probably make these functions `inline(always)` though.